packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4823,6 +4823,13 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
 
                 return oldValue.text != newValue.text || oldValue.composing != newValue.composing;
               },
+              undoStackModifier: (TextEditingValue value) {
+                // On Android we should discard the composing region when pushing
+                // a new entry to the undo stack. This is matches the native
+                // behavior and prevents the TextInputPlugin from resetting the
+                // composing region on every undo/redo.
```

Comment:
It is the TextInputPlugin restarting the input
https://github.com/flutter/engine/blob/cc064aa33ad631807c0b800d132b06e37b193749/shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java#L478-L490 due to the composing region being changed when we do an undo/redo on the framework side.

The Google Keep app has in-app Undo/Redo controls and this is the behavior I observed.

Reply:
> The Google Keep app has in-app Undo/Redo 

I'm being pedantic, but that's not the "native" behavior right? It's just one app that chose to implement it that way. 

> It is the TextInputPlugin restarting the input [here](https://github.com/flutter/engine/blob/cc064aa33ad631807c0b800d132b06e37b193749/shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java#L478-L490) due to the composing region being changed when we do an undo/redo on the framework side.

but the comment is saying this is to prevent the text input plugin from resetting the composing region?

Reply:
Yeah that is true, I have updated the comments to better reflect this and match the pr description.